File: apps/web/lib/diagnostics.ts
import { z } from "zod";
import { apiRequest, csrfToken } from "./auth";

const apiUrl = process.env.NEXT_PUBLIC_CATORA_API_URL ?? "http://localhost:8000";
const uuid = z.string().uuid();

export const DiagnosticStatusSchema = z.enum([
  "awaiting_upload",
  "queued",
  "ingesting",
  "normalizing",
  "categorizing",
  "auditing",
  "matching",
  "preparing_reports",
  "completed",
  "failed",
  "deleting",
]);

export const DiagnosticCountsSchema = z.object({
  processed_rows: z.number().int().nonnegative(),
  accepted_rows: z.number().int().nonnegative(),
  rejected_rows: z.number().int().nonnegative(),
  warning_count: z.number().int().nonnegative(),
  product_count: z.number().int().nonnegative(),
  variant_count: z.number().int().nonnegative(),
  assigned_category_count: z.number().int().nonnegative(),
  ambiguous_category_count: z.number().int().nonnegative(),
  unclassified_category_count: z.number().int().nonnegative(),
  finding_count: z.number().int().nonnegative(),
  intent_run_count: z.number().int().nonnegative(),
  intent_match_count: z.number().int().nonnegative(),
});

export const DiagnosticSchema = z.object({
  id: uuid,
  workspace_id: uuid,
  organization_id: uuid,
  company_name: z.string(),
  status: DiagnosticStatusSchema,
  current_stage: z.string(),
  detail: z.string(),
  market_code: z.string(),
  locale: z.string(),
  currency: z.string(),
  retention_expires_at: z.string(),
  counts: DiagnosticCountsSchema,
  created_at: z.string(),
  updated_at: z.string(),
  completed_at: z.string().nullable(),
  failure_code: z.string().nullable(),
  failure_detail: z.string().nullable(),
  ingestion_job_id: uuid.nullable(),
  audit_run_id: uuid.nullable(),
  intent_run_ids: z.array(uuid),
  result_path: z.string(),
  report_path: z.string(),
  backlog_path: z.string(),
  rejection_path: z.string(),
});

export const DiagnosticRejectionSchema = z.object({
  row_number: z.number().int().positive(),
  reason: z.string(),
  product_handle: z.string().nullable(),
  variant_sku: z.string().nullable(),
});

export const DiagnosticRejectionListSchema = z.object({
  items: z.array(DiagnosticRejectionSchema),
  total_rejected: z.number().int().nonnegative(),
  sample_limit: z.number().int().positive(),
});

export type Diagnostic = z.infer<typeof DiagnosticSchema>;
export type DiagnosticStatus = z.infer<typeof DiagnosticStatusSchema>;
export type DiagnosticRejectionList = z.infer<typeof DiagnosticRejectionListSchema>;

export type CreateDiagnosticInput = {
  company_name: string;
  market_code: string;
  locale: string;
  currency: string;
  retention_days: number;
  authorization_confirmed: boolean;
  storefront_domain?: string;
};

export async function createDiagnostic(
  operatorWorkspaceId: string,
  input: CreateDiagnosticInput,
): Promise<Diagnostic> {
  const payload = await apiRequest<unknown>(
    `/api/v1/workspaces/${operatorWorkspaceId}/prospect-diagnostics`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
  return DiagnosticSchema.parse(payload);
}

export async function uploadDiagnosticCsv(
  assessmentId: string,
  file: File,
): Promise<Diagnostic> {
  const headers = new Headers({
    Accept: "application/json",
    "Content-Type": "text/csv",
  });
  const csrf = csrfToken();
  if (csrf) headers.set("X-CSRF-Token", csrf);
  const response = await fetch(
    `${apiUrl}/api/v1/prospect-diagnostics/${assessmentId}/catalog.csv`,
    {
      method: "PUT",
      headers,
      body: file,
      credentials: "include",
    },
  );
  if (!response.ok) {
    const body = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new Error(body?.detail ?? `Upload failed with status ${response.status}`);
  }
  return DiagnosticSchema.parse(await response.json());
}

export async function getDiagnostic(assessmentId: string): Promise<Diagnostic> {
  const payload = await apiRequest<unknown>(
    `/api/v1/prospect-diagnostics/${assessmentId}`,
  );
  return DiagnosticSchema.parse(payload);
}

export async function getDiagnosticRejections(
  assessmentId: string,
): Promise<DiagnosticRejectionList> {
  const payload = await apiRequest<unknown>(
    `/api/v1/prospect-diagnostics/${assessmentId}/rejections`,
  );
  return DiagnosticRejectionListSchema.parse(payload);
}

export async function deleteDiagnostic(assessmentId: string): Promise<void> {
  await apiRequest<void>(`/api/v1/prospect-diagnostics/${assessmentId}`, {
    method: "DELETE",
  });
}

export function diagnosticDownloadUrl(path: string): string {
  return `${apiUrl}${path}`;
}

export const DIAGNOSTIC_STAGES: ReadonlyArray<{
  status: DiagnosticStatus;
  label: string;
}> = [
  { status: "queued", label: "Queue assessment" },
  { status: "ingesting", label: "Import Shopify catalog" },
  { status: "normalizing", label: "Normalize products and variants" },
  { status: "categorizing", label: "Assign furniture taxonomy" },
  { status: "auditing", label: "Run evidence-backed audit" },
  { status: "matching", label: "Test prepared buyer intents" },
  { status: "preparing_reports", label: "Prepare branded deliverables" },
  { status: "completed", label: "Complete" },
];

export function diagnosticStageState(
  current: DiagnosticStatus,
  stage: DiagnosticStatus,
): "completed" | "active" | "pending" | "failed" {
  if (current === "failed") return "failed";
  const currentIndex = DIAGNOSTIC_STAGES.findIndex((item) => item.status === current);
  const stageIndex = DIAGNOSTIC_STAGES.findIndex((item) => item.status === stage);
  if (current === "completed" || stageIndex < currentIndex) return "completed";
  if (stageIndex === currentIndex) return "active";
  return "pending";
}

File: apps/api/catora_api/schemas/diagnostics.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

DiagnosticStatus = Literal[
    "awaiting_upload",
    "queued",
    "ingesting",
    "normalizing",
    "categorizing",
    "auditing",
    "matching",
    "preparing_reports",
    "completed",
    "failed",
    "deleting",
]


class DiagnosticModel(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)


class DiagnosticCreateRequest(DiagnosticModel):
    company_name: str = Field(min_length=2, max_length=200)
    market_code: str = Field(min_length=2, max_length=35, pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]*$")
    locale: str = Field(min_length=2, max_length=35, pattern=r"^[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*$")
    currency: str = Field(min_length=3, max_length=3, pattern=r"^[A-Za-z]{3}$")
    retention_days: int = Field(default=30, ge=1, le=90)
    authorization_confirmed: bool
    storefront_domain: str | None = Field(default=None, min_length=3, max_length=255)

    @field_validator("company_name")
    @classmethod
    def normalize_company_name(cls, value: str) -> str:
        return " ".join(value.split())

    @field_validator("market_code", "currency")
    @classmethod
    def uppercase_codes(cls, value: str) -> str:
        return value.strip().upper()

    @field_validator("locale")
    @classmethod
    def normalize_locale(cls, value: str) -> str:
        parts = value.strip().split("-")
        return "-".join([parts[0].lower(), *[part.upper() for part in parts[1:]]])

    @field_validator("storefront_domain")
    @classmethod
    def normalize_domain(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip().lower().removeprefix("https://").removeprefix("http://").strip("/")
        if "/" in normalized or " " in normalized or "." not in normalized:
            raise ValueError("storefront_domain must be a hostname")
        return normalized

    @model_validator(mode="after")
    def require_authorization(self) -> DiagnosticCreateRequest:
        if not self.authorization_confirmed:
            raise ValueError("Catalog authorization must be confirmed")
        return self


class DiagnosticCounts(DiagnosticModel):
    processed_rows: int = 0
    accepted_rows: int = 0
    rejected_rows: int = 0
    warning_count: int = 0
    product_count: int = 0
    variant_count: int = 0
    assigned_category_count: int = 0
    ambiguous_category_count: int = 0
    unclassified_category_count: int = 0
    finding_count: int = 0
    intent_run_count: int = 0
    intent_match_count: int = 0


class DiagnosticView(DiagnosticModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    organization_id: uuid.UUID
    company_name: str
    status: DiagnosticStatus
    current_stage: str
    detail: str
    market_code: str
    locale: str
    currency: str
    retention_expires_at: datetime
    counts: DiagnosticCounts
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    failure_code: str | None = None
    failure_detail: str | None = None
    ingestion_job_id: uuid.UUID | None = None
    audit_run_id: uuid.UUID | None = None
    intent_run_ids: list[uuid.UUID] = Field(default_factory=list)
    result_path: str
    report_path: str
    backlog_path: str
    rejection_path: str


class DiagnosticRejection(DiagnosticModel):
    row_number: int = Field(ge=1)
    reason: str
    product_handle: str | None = None
    variant_sku: str | None = None


class DiagnosticRejectionList(DiagnosticModel):
    items: list[DiagnosticRejection]
    total_rejected: int
    sample_limit: int
